core/src/main/java/hudson/util/io/TarArchiver.java

```diff
@@ -87,27 +90,46 @@ public void visit(File file, String relativePath) throws IOException {
         int mode = IOUtils.mode(file);
         if (mode!=-1)   te.setMode(mode);
         te.setModTime(file.lastModified());
-        if(!file.isDirectory())
-            te.setSize(file.length());
-
-        tar.putArchiveEntry(te);
+        long size = 0;
 
         if (!file.isDirectory()) {
-            FileInputStream in = new FileInputStream(file);
-            try {
-                int len;
-                while((len=in.read(buf))>=0)
-                    tar.write(buf,0,len);
-            } finally {
-                in.close();
+            size = file.length();
+            te.setSize(size);
+        }
+        tar.putArchiveEntry(te);
+        try {
+            if (!file.isDirectory()) {
+                FileInputStream in = new FileInputStream(file);
+                try {
+                    int len;
+                    int read = 0;
+                    while ((len = in.read(buf)) >= 0) {
+                        if (len + read <= size) { // ensure we don't write more bytes than the declared when we created
+                                                  // the entry
+                            tar.write(buf, 0, len);
+                            read += len;
+                        } else {
+                            tar.write(buf, 0, (int) size - read);
+                            break;
+                        }
+                    }
+                } catch (IOException e) {// log the exception in any case
+                    LOGGER.log(Level.SEVERE, "Error writing to tar file from: " + file, e);
```

Comment:
I believe the case here was from the fact that if this errors, then the `in.close` 3 lines below in the finally block will almost certainly throw an exception due to an un-closed stream which will then just walk all over this exception

Changing this to a try-with-resources should fix that issue, as well as making the code cleaner, then just go with Jesse's suggestion of catch, wrap and re-throw.